libopenage/datastructure/pairing_heap.h

```diff
@@ -607,18 +546,110 @@ class PairingHeap final {
 	 * O(1)
 	 */
 	void push_node(const element_t &node) {
-		this->root_insert(node);
+		auto [iter, result] = this->nodes.insert(node);
+		if (result) {
+			this->root_insert(node);
+			this->node_count += 1;
+		}
 
 #if OPENAGE_PAIRINGHEAP_DEBUG
-		auto ins = this->nodes.insert(node);
-		if (not ins.second) {
+		if (not result) {
 			throw Error{ERR << "node already known"};
 		}
 #endif
+	}
+
+
+	/**
+	 * returns the smallest item on the heap and deletes it.
+	 * also known as delete_min.
+	 *                       _________
+	 * Ω(log log n), O(2^(2*√log log n'))
+	 * caller must eventually either add node back to heap or delete it
+	 */
+	element_t pop_node() {
```

Comment:
It looks like `pop_node()` no longer deletes the node from the pointer, so this may cause issues. Especially if other methods like `unlink_node()` expect that `pop_node()` does deletion.

If you look back at https://github.com/SFTtech/openage/commit/4ab602a6242448bb82eac5db9eafa96d758c7199#diff-84a922def1ded03f0d8849cb1c1d56e758f9ae47a06cb78fc40b193294d40ab9R380 where the shared:ptr implementation was introduced, you can see that there originally was no `pop_node()` (only `pop()`). Maybe it makes sense to merge the methods again.

Reply:
I noticed as well that's, why I made `pop_node()` private, I thought maybe as long as its callers handle the node correctly, it should be fine. I will just merge pop and pop_node.

Reply:
In the old implementation, `pop_node` was technically only a delete (more akin to `remove_node` now). It's a bit confusing bcause of the renames. Merging the functions is the right call I think.